libpam/include/security/pam_sandbox.h

```diff
@@ -0,0 +1,136 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/landlock.h>
+#include <linux/prctl.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/prctl.h>
+#include <sys/stat.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#ifndef landlock_create_ruleset
+
+	static inline int
+landlock_create_ruleset(const struct landlock_ruleset_attr *const attr,
+		const size_t size, const __u32 flags)
+{
+	return syscall(__NR_landlock_create_ruleset, attr, size, flags);
+}
+#endif
+
+#ifndef landlock_add_rule
+static inline int landlock_add_rule(const int ruleset_fd,
+		const enum landlock_rule_type rule_type,
+		const void *const rule_attr,
+		const __u32 flags)
+{
+	return syscall(__NR_landlock_add_rule, ruleset_fd, rule_type, rule_attr,
+			flags);
+}
+#endif
+
+#ifndef landlock_restrict_self
+static inline int landlock_restrict_self(const int ruleset_fd,
+		const __u32 flags)
+{
+	return syscall(__NR_landlock_restrict_self, ruleset_fd, flags);
+}
+#endif
+
+#define ACCESS_FS_ROUGHLY_READ ( \
+		LANDLOCK_ACCESS_FS_EXECUTE | \
+		LANDLOCK_ACCESS_FS_READ_FILE | \
+		LANDLOCK_ACCESS_FS_READ_DIR)
+
+#ifdef LANDLOCK_ACCESS_FS_REFER
+#define ACCESS_FS_ROUGHLY_WRITE ( \
+		LANDLOCK_ACCESS_FS_WRITE_FILE | \
+		LANDLOCK_ACCESS_FS_REMOVE_DIR | \
+		LANDLOCK_ACCESS_FS_REMOVE_FILE | \
+		LANDLOCK_ACCESS_FS_MAKE_CHAR | \
+		LANDLOCK_ACCESS_FS_MAKE_DIR | \
+		LANDLOCK_ACCESS_FS_MAKE_REG | \
+		LANDLOCK_ACCESS_FS_MAKE_SOCK | \
+		LANDLOCK_ACCESS_FS_MAKE_FIFO | \
+		LANDLOCK_ACCESS_FS_MAKE_BLOCK | \
+		LANDLOCK_ACCESS_FS_MAKE_SYM | \
+		LANDLOCK_ACCESS_FS_REFER)
+
+#define ACCESS_ABI_2 ( \
+		LANDLOCK_ACCESS_FS_REFER)
+#else
+#define ACCESS_FS_ROUGHLY_WRITE ( \
+		LANDLOCK_ACCESS_FS_WRITE_FILE | \
+		LANDLOCK_ACCESS_FS_REMOVE_DIR | \
+		LANDLOCK_ACCESS_FS_REMOVE_FILE | \
+		LANDLOCK_ACCESS_FS_MAKE_CHAR | \
+		LANDLOCK_ACCESS_FS_MAKE_DIR | \
+		LANDLOCK_ACCESS_FS_MAKE_REG | \
+		LANDLOCK_ACCESS_FS_MAKE_SOCK | \
+		LANDLOCK_ACCESS_FS_MAKE_FIFO | \
+		LANDLOCK_ACCESS_FS_MAKE_BLOCK | \
+		LANDLOCK_ACCESS_FS_MAKE_SYM)
+#endif
+
+#define ACCESS_FILE_RW ( \
+		LANDLOCK_ACCESS_FS_WRITE_FILE | \
+		LANDLOCK_ACCESS_FS_READ_FILE)
+
+#define ACCESS_FILE_RO (LANDLOCK_ACCESS_FS_READ_FILE)
+
+
+static int _restrict_to_file_read(const char * const file_path)
+{
+	__u64 access_fs_rw = ACCESS_FS_ROUGHLY_READ | ACCESS_FS_ROUGHLY_WRITE;
+
+	struct landlock_ruleset_attr ruleset_attr = {
+		.handled_access_fs = access_fs_rw,
+	};
+	struct landlock_path_beneath_attr path_beneath = {
+		.allowed_access = ACCESS_FILE_RO,
+	};
+	int ruleset_fd, err;
+#ifdef LANDLOCK_ACCESS_FS_REFER
```

Comment:
`LANDLOCK_ACCESS_FS_REFER` may be defined or not by the libc headers, but what matter is what the running kernel supports. See the Suricata patches which defines `LANDLOCK_ACCESS_FS_REFER` if not already defined.

In a general case, this kind of ABI check should be performed all the time to make the code less complex and more maintainable.

Anyway, as said just before, for now there is no need to handle `LANDLOCK_ACCESS_FS_REFER`. This kind of ABI version check will be useful when Landlock will handle more interesting access rights (e.g. the `LANDLOCK_ACCESS_FS_TRUNCATE` one). The only thing required for now is to check that the running kernel support Landlock (i.e. `if (abi < 0) { return without error }`).

Reply:
Thank's I now use the suricata way to handle this.